crates/matrix-sdk-base/src/rooms/calls.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Things related to the room calls (also known as MatrixRTC sessions).
+
+use std::{collections::HashMap, time::Duration};
+
+use ruma::{
+    events::call::member::SyncCallMemberEvent, MilliSecondsSinceUnixEpoch, OwnedDeviceId,
+    OwnedUserId,
+};
+use serde::{Deserialize, Serialize};
+
+/// Stores and handles the information related to the memberships of the calls
+/// in a room.
+///
+/// Currently we only assume that we support `m.call` application types of the
+/// calls and only the room calls (one such call per room).
+#[derive(Debug, Default, Deserialize, Clone, Serialize)]
+pub(crate) struct CallMembershipsHandler {
+    /// Members of the call. We only store members of the active room call.
+    memberships: HashMap<OwnedUserId, Vec<CallMember>>,
+}
+
+impl CallMembershipsHandler {
+    /// Processes incoming sync call member event.
+    pub(crate) fn handle_call_member_event(&mut self, event: &SyncCallMemberEvent) {
+        let Some(event) = event.as_original().cloned() else {
+            return;
+        };
+
+        let user_id = event.state_key;
+        let user_members =
+            event.content.memberships.into_iter().filter(|m| m.is_room_call()).map(|m| {
+                CallMember {
+                    device_id: m.device_id.into(),
+                    info: CallMemberInfo {
+                        member_since: m.created_ts.unwrap_or(event.origin_server_ts),
+                        valid_for: m.expires,
+                    },
+                }
```

Comment:
We, unfortunately, cannot define `From<Membership> for CallMember`, because we need access to the original `event.origin_server_ts` to construct a `CallMember`. This is essentially the whole point why I do not use the "raw" `ruma` event here because it represents a different entity (an unverified raw event with possibly missing start timestamp).